File: app-flutter/lib/features/auth/service/auth_service.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart' show PlatformException;
import 'package:google_sign_in/google_sign_in.dart';
import 'package:nebras_mobile_app/features/auth/model/user_profile.dart';

/// واجهة رفيعة تلفّ `google_sign_in` و `firebase_auth` لعزل بقيّة التطبيق
/// عن تفاصيل الحزمة.
///
/// Android: `applicationId` يختلف بين `flutter run` و release — راجع
/// [docs/GOOGLE_SIGNIN_ANDROID.md](../../../docs/GOOGLE_SIGNIN_ANDROID.md).
///
/// تصميم تدفّق الدخول:
///   1. `signInSilently()` عند الإقلاع — يستعيد الحساب إن كان المستخدم
///       قد سجّل دخوله سابقاً، ويعيد `null` بسلاسة إن لم يفعل.
///   2. `signInInteractive()` عند ضغط المستخدم على زرّ "الدخول بـ Google"
///       — يفتح نافذة اختيار الحساب الخاصّة بنظام Android ويسجل الدخول في Firebase.
class AuthService {
  AuthService()
    : _googleSignIn = GoogleSignIn(
        scopes: const ['email', 'profile', 'openid'],
        // Web client (OAuth type 3) — مطلوب لـ Firebase Auth idToken على Android.
        serverClientId:
            '412379996427-0i55p23iunk93tilvje5442d6e42s4p9.apps.googleusercontent.com',
      ),
      _auth = FirebaseAuth.instance;

  final GoogleSignIn _googleSignIn;
  final FirebaseAuth _auth;

  /// محاولة دخول صامتة — لا تُظهر أيّ واجهة. تُستخدم أثناء الإقلاع.
  Future<UserProfile?> signInSilently() async {
    try {
      final account = await _googleSignIn.signInSilently();
      if (account == null) return null;

      return await _firebaseAuthWithGoogle(account);
    } catch (e, st) {
      debugPrint('[AuthService] signInSilently failed: $e\n$st');
      return null;
    }
  }

  /// دخول تفاعلي — يفتح مُنتقي حسابات Google على الجهاز ويسجل في Firebase.
  Future<UserProfile?> signInInteractive() async {
    try {
      final account = await _googleSignIn.signIn();
      if (account == null) return null;

      return await _firebaseAuthWithGoogle(account);
    } on PlatformException catch (e) {
      debugPrint(
        '[AuthService] signIn PlatformException: ${e.code} ${e.message}',
      );
      throw AuthException(
        code: e.code,
        message: e.message ?? 'Google sign-in failed',
      );
    } catch (e, st) {
      debugPrint('[AuthService] signIn unknown error: $e\n$st');
      throw AuthException(code: 'unknown', message: e.toString());
    }
  }

  Future<UserProfile?> _firebaseAuthWithGoogle(
    GoogleSignInAccount account,
  ) async {
    final GoogleSignInAuthentication googleAuth = await account.authentication;

    final AuthCredential credential = GoogleAuthProvider.credential(
      accessToken: googleAuth.accessToken,
      idToken: googleAuth.idToken,
    );

    final UserCredential userCredential = await _auth.signInWithCredential(
      credential,
    );
    return _mapAccount(userCredential.user);
  }

  Future<void> signOut() async {
    try {
      await _auth.signOut();
      await _googleSignIn.signOut();
    } catch (e) {
      debugPrint('[AuthService] signOut error: $e');
    }
  }

  /// حذف الحساب نهائيّاً من Firebase Authentication. إن طلب Firebase
  /// إعادة مصادقة حديثة (`requires-recent-login`) نُعيد المصادقة بـ Google
  /// ثمّ نحذف. في كلّ الأحوال نُسجّل الخروج من Google بعدها.
  Future<void> deleteAccount() async {
    final user = _auth.currentUser;
    if (user == null) return;
    try {
      await user.delete();
    } on FirebaseAuthException catch (e) {
      if (e.code == 'requires-recent-login') {
        final account = await _googleSignIn.signIn();
        if (account == null) {
          throw const AuthException(
            code: 'reauth_cancelled',
            message: 'Re-authentication required to delete the account',
          );
        }
        final googleAuth = await account.authentication;
        final credential = GoogleAuthProvider.credential(
          accessToken: googleAuth.accessToken,
          idToken: googleAuth.idToken,
        );
        await user.reauthenticateWithCredential(credential);
        await user.delete();
      } else {
        throw AuthException(code: e.code, message: e.message ?? 'Delete failed');
      }
    } finally {
      try {
        await _googleSignIn.signOut();
      } catch (_) {}
    }
  }

  Future<void> disconnect() async {
    try {
      await _auth.signOut();
      await _googleSignIn.disconnect();
    } catch (_) {
      /* قد يرمي إن كان المستخدم مسجّلاً أصلاً — نتجاهل بأمان. */
    }
  }

  UserProfile? _mapAccount(User? user) {
    if (user == null) return null;
    return UserProfile(
      id: user.uid,
      email: user.email ?? '',
      displayName: user.displayName?.trim().isNotEmpty == true
          ? user.displayName!
          : (user.email != null ? user.email!.split('@').first : 'User'),
      photoUrl: user.photoURL,
      signedInAt: DateTime.now(),
    );
  }
}

class AuthException implements Exception {
  final String code;
  final String message;
  const AuthException({required this.code, required this.message});

  @override
  String toString() => 'AuthException($code): $message';
}

File: nebras-app/app/src/main/kotlin/com/nebras/mobile/feature/auth/AuthService.kt
package com.nebras.mobile.feature.auth

import android.content.Context
import android.util.Log
import androidx.credentials.ClearCredentialStateRequest
import androidx.credentials.CredentialManager
import androidx.credentials.CustomCredential
import androidx.credentials.GetCredentialRequest
import androidx.credentials.exceptions.GetCredentialCancellationException
import androidx.credentials.exceptions.NoCredentialException
import com.google.android.libraries.identity.googleid.GetGoogleIdOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthRecentLoginRequiredException
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GoogleAuthProvider
import kotlinx.coroutines.tasks.await
import org.json.JSONObject

/** ملفّ المستخدم — نقل `auth/model/user_profile.dart`. */
data class UserProfile(
    val id: String,
    val email: String,
    val displayName: String,
    val photoUrl: String? = null,
    val signedInAtMs: Long,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id)
        put("email", email)
        put("displayName", displayName)
        photoUrl?.let { put("photoUrl", it) }
        put("signedInAtMs", signedInAtMs)
    }

    companion object {
        fun fromJson(json: JSONObject): UserProfile = UserProfile(
            id = json.optString("id"),
            email = json.optString("email"),
            displayName = json.optString("displayName"),
            photoUrl = json.optString("photoUrl").takeIf { it.isNotEmpty() },
            signedInAtMs = json.optLong("signedInAtMs", System.currentTimeMillis()),
        )
    }
}

class AuthException(val code: String, message: String) : Exception(message) {
    override fun toString(): String = "AuthException($code): $message"
}

/**
 * خدمة تسجيل الدخول بـ Google — نقل `auth/service/auth_service.dart`.
 *
 * بديل حزمة `google_sign_in`: **Credential Manager** + `googleid` (الطريق
 * المعتمد حالياً من Google). نفس `serverClientId` المسجَّل في Firebase.
 */
class AuthService(private val context: Context) {

    private companion object {
        const val TAG = "AuthService"

        /** معرّف عميل الويب من Firebase — نفس قيمة نسخة Flutter حرفياً. */
        const val SERVER_CLIENT_ID =
            "412379996427-0i55p23iunk93tilvje5442d6e42s4p9.apps.googleusercontent.com"
    }

    private val credentialManager = CredentialManager.create(context)
    private val auth = FirebaseAuth.getInstance()

    /**
     * دخول صامت: حساب Google مصرَّح سابقاً فقط (بلا واجهة). يعود null عند
     * عدم وجود اعتماد — نفس دلالة `signInSilently`.
     */
    suspend fun signInSilently(): UserProfile? = runCatching {
        // مستخدم Firebase حيّ من جلسة سابقة يكفي — لا حاجة لحوار.
        auth.currentUser?.let { return mapUser(it) }

        val option = GetGoogleIdOption.Builder()
            .setServerClientId(SERVER_CLIENT_ID)
            .setFilterByAuthorizedAccounts(true)
            .setAutoSelectEnabled(true)
            .build()
        val response = credentialManager.getCredential(
            context,
            GetCredentialRequest.Builder().addCredentialOption(option).build(),
        )
        firebaseAuthWithCredential(response.credential)
    }.getOrElse {
        if (it !is NoCredentialException) {
            Log.d(TAG, "signInSilently failed: ${it.message}")
        }
        null
    }

    /**
     * دخول تفاعليّ من [activityContext] (حوار اختيار الحساب يحتاج Activity).
     * يعود null إن ألغى المستخدم.
     */
    suspend fun signInInteractive(activityContext: Context): UserProfile? {
        try {
            val option = GetGoogleIdOption.Builder()
                .setServerClientId(SERVER_CLIENT_ID)
                .setFilterByAuthorizedAccounts(false)
                .build()
            val response = credentialManager.getCredential(
                activityContext,
                GetCredentialRequest.Builder().addCredentialOption(option).build(),
            )
            return firebaseAuthWithCredential(response.credential)
        } catch (e: GetCredentialCancellationException) {
            return null // ألغى المستخدم.
        } catch (e: NoCredentialException) {
            throw AuthException("sign_in_failed", e.message ?: "Google sign-in failed")
        } catch (e: Throwable) {
            Log.d(TAG, "signIn unknown error: ${e.message}")
            throw AuthException("unknown", e.toString())
        }
    }

    private suspend fun firebaseAuthWithCredential(
        credential: androidx.credentials.Credential,
    ): UserProfile? {
        if (credential !is CustomCredential ||
            credential.type != GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL
        ) {
            throw AuthException("sign_in_failed", "Unexpected credential type")
        }
        val idToken = GoogleIdTokenCredential.createFrom(credential.data).idToken
        val firebaseCredential = GoogleAuthProvider.getCredential(idToken, null)
        val result = auth.signInWithCredential(firebaseCredential).await()
        return mapUser(result.user)
    }

    suspend fun signOut() {
        runCatching {
            auth.signOut()
            credentialManager.clearCredentialState(ClearCredentialStateRequest())
        }.onFailure { Log.d(TAG, "signOut error: ${it.message}") }
    }

    /**
     * حذف الحساب من Firebase Auth. عند `requires-recent-login` نُعيد
     * المصادقة تفاعليّاً ثمّ نحذف — نفس مسار نسخة Flutter.
     */
    suspend fun deleteAccount(activityContext: Context) {
        val user = auth.currentUser ?: return
        try {
            user.delete().await()
        } catch (e: FirebaseAuthRecentLoginRequiredException) {
            val option = GetGoogleIdOption.Builder()
                .setServerClientId(SERVER_CLIENT_ID)
                .setFilterByAuthorizedAccounts(false)
                .build()
            val response = try {
                credentialManager.getCredential(
                    activityContext,
                    GetCredentialRequest.Builder().addCredentialOption(option).build(),
                )
            } catch (cancel: GetCredentialCancellationException) {
                throw AuthException(
                    "reauth_cancelled",
                    "Re-authentication required to delete the account",
                )
            }
            val credential = response.credential
            if (credential is CustomCredential &&
                credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL
            ) {
                val idToken = GoogleIdTokenCredential.createFrom(credential.data).idToken
                user.reauthenticate(GoogleAuthProvider.getCredential(idToken, null)).await()
                user.delete().await()
            } else {
                throw AuthException("sign_in_failed", "Unexpected credential type")
            }
        } finally {
            runCatching {
                credentialManager.clearCredentialState(ClearCredentialStateRequest())
            }
        }
    }

    suspend fun disconnect() {
        runCatching {
            auth.signOut()
            credentialManager.clearCredentialState(ClearCredentialStateRequest())
        } // قد يرمي إن كان المستخدم خارجاً أصلاً — نتجاهل بأمان.
    }

    private fun mapUser(user: FirebaseUser?): UserProfile? {
        if (user == null) return null
        return UserProfile(
            id = user.uid,
            email = user.email.orEmpty(),
            displayName = user.displayName?.trim()?.takeIf { it.isNotEmpty() }
                ?: user.email?.substringBefore('@')
                ?: "User",
            photoUrl = user.photoUrl?.toString(),
            signedInAtMs = System.currentTimeMillis(),
        )
    }
}
